docs/optimistic_locking.rst

```diff
@@ -0,0 +1,175 @@
+          host = "http://localhost:8000"
+      office_id = UnicodeAttribute(hash_key=True)
+      employees = ListAttribute(of=OfficeEmployeeMap)
+      name = UnicodeAttribute()
+      version = VersionAttribute()
+
+The attribute is underpinned by an integer which is initialized with 1 when an item is saved for the first time
+and is incremented by 1 with each subsequent write operation.
+
+.. code-block:: python
+
+  justin = OfficeEmployeeMap(office_employee_id=str(uuid4()), person='justin')
+  garrett = OfficeEmployeeMap(office_employee_id=str(uuid4()), person='garrett')
+  office = Office(office_id=str(uuid4()), name="office", employees=[justin, garrett])
+  office.save()
+  assert office.version == 1
+
+  # Get a second local copy of Office
+  office_out_of_date = Office.get(office.office_id)
+
+  # Add another employee and persist the change.
+  office.employees.append(OfficeEmployeeMap(office_employee_id=str(uuid4()), person='lita'))
+  office.save()
+  # On subsequent save or update operations the version is also incremented locally to match the persisted value so
+  # there's no need to refresh between operations when reusing the local copy.
+  assert office.version == 2
+  assert office_out_of_date.version == 1
+
+The version checking is implemented using DynamoDB conditional write constraints. Asserting that no value exists
+for the version attribute on the initial save and that the persisted value matches the local value on subsequent writes.
+
+
+Model.{update, save, delete}
+----------------------------
+These operations will fail if the local object is out-of-date.
+
+.. code-block:: python
+
+  @contextmanager
+  def assert_condition_check_fails():
+      try:
+          yield
+      except (PutError, UpdateError, DeleteError) as e:
+          assert isinstance(e.cause, ClientError)
+          assert e.cause_response_code == "ConditionalCheckFailedException"
+      except TransactWriteError as e:
+          assert isinstance(e.cause, ClientError)
+          assert e.cause_response_code == "TransactionCanceledException"
+          assert "ConditionalCheckFailed" in e.cause_response_message
+      else:
+          raise AssertionError("The version attribute conditional check should have failed.")
+
+
+  with assert_condition_check_fails():
+      office_out_of_date.update(actions=[Office.name.set('new office name')])
+
+  office_out_of_date.employees.remove(garrett)
+  with assert_condition_check_fails():
+      office_out_of_date.save()
+
+  # After refreshing the local copy our write operations succeed.
+  office_out_of_date.refresh()
+  office_out_of_date.employees.remove(garrett)
+  office_out_of_date.save()
+  assert office_out_of_date.version == 3
+
+  with assert_condition_check_fails():
+      office.delete()
+
+Transactions
+------------
+
+Transactions are supported.
+
+**Note:** *local objects that were used in a successful write transaction will not have their version attribute updated
```

Comment:
Is this something we should fix instead of including a warning?

Reply:
Fixed, updated tests + example + docs